### PR TITLE
Perspective UI prototype

### PR DESCRIPTION
**Pull request by Gozala**
For now I primarily added `type` directory that encodes details present in the specification diagram into type signatures. From this point on I intend to build up a UI that fulfills presented types signatures.

### CODE REVIEW

File: type/browser/shell.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/* @flow */
+
+import * as Focusable from "../common/focusable"
+import * as Hoverable from "../common/focusable"
```
Comment (Gozala) [inline type/browser/shell.js]: should be '../common/hoverable'

File: type/common/focusable.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+export type Action
+  = RequestFocus
+  | Focus
+  | Blur
+
+type Focusable = Model
+
+
+export type focus <Model:Focusable> = (model:Model) => Model
```
Comment (Gozala) [inline type/common/focusable.js]:
```js
type X = { isX: true } & Focusable
type Y = { isY: true } & Focusable

type foo = (model:Focusable) => Focusable
```

File: type/browser/input.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import * as Focusable from "../common/focusable"
+
+export type Model
+  = Editable.Model
+  & Focusable.Model
+
+
+export type Action
+  = Editable.Action
+  & Focusable.Action
```
Comment (Gozala) [inline type/browser/input.js]: should be `|` instead of `&`

File: type/browser/browser.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  | For<"Assistant", Assistant.Action>
+  | For<"Devtools", Devtools.Action>
+
+export type Response
+  = For<"WebViews", WebViewList.Action>
+  | For<"Assistant", Assistant.Response>
+  | For<"Devtools", Devtools.Response>
+
+export type view = (model:Model, address:Address<Action>) => VirtualTree
+export type step = (model:Model, action:Action) => [Model, Effects<Response>]
```
Comment (Gozala) [inline type/browser/browser.js]:
```js
const asFor = target => action => ({type: "For", target, action})
const merge = (target, extra) => Object.assign({}, target, extra)

view = (model, address) =>
  html.div({
    key: 'root',
    title: //...,
    onFocus: onWindow(forward(address, asFor("Shell")), Focusable.asFocus),
    onBlur: onWindow(forward(address, asFor("Blur")), Focusable.asBlur),
  }, [
    WindowControls.view(...),
    Devtools.view(model.devtools, forward(address, asFor("Devtools"))),
    // ...
  ])


step = (model, action) => {
  if (action.type === "For") {
    if (action.target === "Shell") {
      return [
        merge(model, {
          shell: Shell.update(model.shell, action.action)
        }),
        Effects.none
      ]
    }

     if (action.target === "Devtools") {
       const [devtools, fx] = Devtools.step(model.devtools, action.action)
       return [
         merge(model, {devtools}),
         fx.map(asFor("Devtools"))
       ]
     }
  }
}
```

**Comment by gordonbrander** [comment]
We previously had higher-order functions like `on` and `onWindow` https://github.com/mozilla/browser.html/blob/1e04184e149a299df2cc8008dc1433ed6a04120e/src/driver/virtual-dom/index.js#L120. These seem to have disappeared.

**Comment by gordonbrander** [comment]
Do we use immutable.js at all in this new version? It doesn't look like it.

**Comment by gordonbrander** [comment]
FYI: I'm attempting to build on top of this PR here https://github.com/Gozala/browser.html/pull/2/files

**Comment by gordonbrander** [comment]
Updated PR here https://github.com/mozilla/browser.html/pull/768

**Pull request closed without merging**